src/Tools/AI Test Toolkit/src/Logs/AITLogEntry.Table.al

```diff
@@ -0,0 +1,282 @@
+            IncludedFields = "Procedure Name", Status;
+        }
+        key(Key4; "Test Suite Code", Version, Operation, "Duration (ms)")
+        {
+            // Instead of a SIFT index. This will make both inserts and calculations faster - and non-blocking
+        }
+        key(Key3; "Duration (ms)")
+        {
+            SumIndexFields = "Duration (ms)";
+        }
+    }
+
+
+    procedure SetInputBlob(NewInput: Text)
+    var
+        OutStream: OutStream;
+    begin
+        Clear("Input Data");
+        "Input Data".CreateOutStream(OutStream, this.GetDefaultTextEncoding());
+        OutStream.Write(NewInput);
+    end;
+
+    procedure GetInputBlob(): Text
+    var
+        InStream: InStream;
+        InputContent: Text;
+    begin
+        this.CalcFields("Input Data");
+        "Input Data".CreateInStream(InStream, this.GetDefaultTextEncoding());
+        InStream.Read(InputContent);
+        exit(InputContent);
+    end;
+
+    procedure SetOutputBlob(NewOutput: Text)
+    var
+        OutStream: OutStream;
+    begin
+        Clear("Output Data");
+        "Output Data".CreateOutStream(OutStream, this.GetDefaultTextEncoding());
+        OutStream.Write(NewOutput);
+    end;
+
+    procedure GetOutputBlob(): Text
+    var
+        InStream: InStream;
+        OutputContent: Text;
+    begin
+        this.CalcFields("Output Data");
+        "Output Data".CreateInStream(InStream, this.GetDefaultTextEncoding());
+        InStream.Read(OutputContent);
+        exit(OutputContent);
+    end;
+
+    procedure SetMessage(Msg: Text)
+    var
+        MessageOutStream: OutStream;
+    begin
+        Clear(Message);
+        Message.CreateOutStream(MessageOutStream, this.GetDefaultTextEncoding());
+        MessageOutStream.WriteText(Msg);
+    end;
+
+    procedure GetMessage(): Text
+    var
+        MessageInStream: InStream;
+        MessageText: Text;
+    begin
+        this.CalcFields(Message);
+        Message.CreateInStream(MessageInStream, this.GetDefaultTextEncoding());
+        MessageInStream.ReadText(MessageText);
+        exit(MessageText);
+    end;
+
+    procedure SetErrorCallStack(ErrorCallStack: Text)
+    var
+        ErrorCallStackOutStream: OutStream;
+    begin
+        Clear("Error Call Stack");
+        "Error Call Stack".CreateOutStream(ErrorCallStackOutStream, this.GetDefaultTextEncoding());
+        ErrorCallStackOutStream.WriteText(ErrorCallStack);
+    end;
+
+    procedure GetErrorCallStack(): Text
+    var
+        ErrorCallStackInStream: InStream;
+        ErrorCallStackText: Text;
+    begin
+        this.CalcFields("Error Call Stack");
+        "Error Call Stack".CreateInStream(ErrorCallStackInStream, this.GetDefaultTextEncoding());
+        ErrorCallStackInStream.ReadText(ErrorCallStackText);
+        exit(ErrorCallStackText);
+    end;
+
+    local procedure GetDefaultTextEncoding(): TextEncoding
+    begin
+        exit(TextEncoding::UTF8);
+    end;
+
+    trigger OnInsert()
```

Comment:
Please put the trigger ontop of the functions.

Reply:
@darjoo 
A general question: I know - and use - that pattern to first declare the triggers, and then other procedures, but this is actually nowhere documented, or I cannot find it. If there are more AL rules Microsoft-written objects follow, it would be really helpful to have them listed all in one place. Or could you point to me to such a website?

Reply:
To be honest, I don't know if it is documented anywhere. It's a soft code style we try to follow internally.

Reply:
We have no hard rule or guideline other than a wish to make the code look consistent. Here it is legacy based on what code looked like in C/SIDE where each element had a fixed position in the file which roughly went like this:
global properties, type specific elements (page controls, data items etc), then triggers, globals and finally user defined functions.
The .al format is much more flexible in these things (and of course we no longer run the code through an import/export process. Of course the Txt2Al tool converted the C/AL to al keeping many of these things in the same position which is why most BaseApp code still has the same old order.

Reply:
now I've made myself wonder... did C/AL put triggers before or after globals? hmmm

Reply:
Oh wow... my memory needed a refresher :P
so it was (for a page example):
properties, triggers, actions, controls, globals, procedures

in .al I think we got:
properties, controls, actions, triggers, globals, procedures